[FSDP] Add grad accumulation without `no_sync()`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73787 [FSDP][BE] Change assert to assertEqual
* #73786 [Easy][FSDP] Fix warning render
* **#73535 [FSDP] Add grad accumulation without `no_sync()`**

**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

Commit:
[FSDP] Add grad accumulation without `no_sync()`

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.



**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.

This also adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1305,7 +1305,18 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
+                    )
+                    param._saved_grad_shard.data += output.data  # type: ignore[attr-defined]
+                else:
+                    param._saved_grad_shard = output.data  # type: ignore[attr-defined]
```
are we switching from using `output` to `output.data` in non-grad accumulation use case?

awgu replied:
Good question. This is good for me to clarify. Was there any reason to use `output` before?

In my understanding, it does not matter since we are not supporting taking the gradient of the gradient, so we do not need to record any operations on `output` in the autograd graph. Fairscale uses `output.data` in both code paths, though I do not see why either way is better for the non-gradient accumulation case.

rohan-varma commented on test/distributed/fsdp/test_fsdp_grad_acc.py:
```diff
@@ -132,8 +162,13 @@ def permute_tensor(x: torch.Tensor):
-        [2, 4],
+        "configs",
+        [
+            [_GradAccConfig(True, 4)],
```
nit: if you want, maybe pass in named args here, so future developers know what it is

rohan-varma commented on test/distributed/fsdp/test_fsdp_grad_acc.py:
```diff
@@ -132,8 +162,13 @@ def permute_tensor(x: torch.Tensor):
+            [_GradAccConfig(True, 4)],
+            [_GradAccConfig(False, 4)],
+            [_GradAccConfig(True, 2), _GradAccConfig(False, 2), _GradAccConfig(True, 2)],
+            [_GradAccConfig(False, 2), _GradAccConfig(True, 2), _GradAccConfig(False, 2)],
```
Wondering why we have duplicated configs?

awgu replied:
I wanted to test interleaving both ways:
1) with `no_sync()` -> without `no_sync()` -> with `no_sync()`
2) without `no_sync()` -> with `no_sync()` -> without `no_sync()`

The reason I wanted two separate tests is that it _could_ matter what was the last accumulation mode right before gradient synchronization.

rohan-varma replied:
Good call! Let's keep it then and add a small comment so no one removes it in the future.

rohan-varma commented on test/distributed/fsdp/test_fsdp_grad_acc.py:
```diff
@@ -132,8 +162,13 @@ def permute_tensor(x: torch.Tensor):
-        [2, 4],
+        "configs",
+        [
+            [_GradAccConfig(True, 4)],
```
to reduce # of tests, can we just have:

use_context = true, interval={2,4}
use_context = false, interval={2,4}

awgu replied:
Given your comments last time, I thought about how to keep the set of tests minimal. I feel like these 4 configs test distinct things:
1) Gradient accumulation with `no_sync()`
2) Gradient accumulation without `no_sync()`
3) Gradient accumulation interleaving `no_sync()` and without `no_sync()`, where the last iteration before synchronizing gradients is in `no_sync()`
4) Gradient accumulation interleaving `no_sync()` and without `no_sync()`, where the last iteration before synchronizing gradients is outside `no_sync()`

My concern is that I do not want to overfit the current working implementation. It not obvious to me that 1) and 2) imply 3) and 4), and if we only had 3) and 4) and they break, then we would probably end up rewriting 1) and 2).

rohan-varma replied:
Sounds good! Considering you've gone through the tradeoff this makes sense to leave as is

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1305,7 +1305,18 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                     # Average grad by world_size for consistency with PyTorch DDP.
                     output.div_(self.gradient_postdivide_factor)
-                param.grad.data = output
+                accumulate_grad = getattr(param, "_saved_grad_shard", None) is not None
```
Might be useful to add a small comment about how gradient accumulation without no_sync is implemented. From my understanding:

1. During backward, we point an attribute _saved_grad_shard to the gradient shard
2. If we are accumulating gradients, we accumulate it on _saved_grad_shard
3. When finalizing backwards before running the optimizer, we point p.grad to the saved grad shard so that the optimizer works on the right accumulated gradient.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1485,6 +1510,10 @@ def _prep_grads_for_backward(self) -> None:
+                can_accumulate_grad = p.grad.device == p.data.device and \
+                    p.grad.size() == p._local_shard.shape  # type: ignore[attr-defined]
+                if can_accumulate_grad:
+                    p._saved_grad_shard = p.grad.data  # type: ignore[attr-defined]
```
add comment here that this is what makes the gradient accumulation work?

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1485,6 +1510,10 @@ def _prep_grads_for_backward(self) -> None:
                 p.grad.size() != p._orig_size  # type: ignore[attr-defined]
                 or p.grad.device != p.device
             ):
+                can_accumulate_grad = p.grad.device == p.data.device and \
```
Can we have a unittest where can_accumulate_grad = False and the user tries to accumulate grads and we raise an appropriate error?

awgu replied:
After some investigation, I think that having a non-silent error for the case of gradient accumulation outside `no_sync()` while using CPU offloading requires some non-trivial re-design. It may be easiest to leave this error as silent for now and work on adding the compatibility itself.

(_Some clarifying questions and comments_)
Suppose we are using CPU offloading.
- Outside `no_sync()`, if we want to accumulate gradients, are we performing the addition between the existing gradient and the newly-reduced gradient on CPU or on GPU? If on CPU, then we should perform a device-to-host transfer of the reduced gradient. If on GPU, then we should perform a host-to-device transfer of the existing gradient and a device-to-host transfer to re-offload the result.
- Inside `no_sync()`, the existing implementation does not offload any gradients to CPU. Rather, the gradients are held in GPU memory until the first iteration outside `no_sync()`, which performs the gradient synchronization. At the end of that iteration's backward pass, the synchronized gradient shard is offloaded to CPU. Should we include any warning or message at runtime to the explain this behavior to the user? I added a note about it to the `no_sync()` docstring.

awgu replied:
The challenge behind a non-silent error is that the `no_sync()` + CPU offloading case conflicts with the non-`no_sync()` + CPU offloading case.

- The crux is that accumulating gradients using `no_sync()` follows the pattern of: accumulate for `N-1` iterations inside `no_sync()` and **execute `1` normal iteration outside `no_sync()`**.
- That final normal iteration is **indistinguishable** from a non-`no_sync()` iteration unless we track something like a  `bool` flag indicating that the last iteration was inside `no_sync()`.
    - I am reluctant to add such a flag since it is solely a patch for one case and may be hiding the underlying design problem, but I am open to your thoughts.
- For the final iteration coming out of `no_sync()`, the gradient is still on GPU, so performing the accumulation computation `param._saved_grad_shard.data += output.data` has no issue.
- For a non-`no_sync()` iteration (after the first), the gradient was previously offloaded to CPU, so performing the accumulation computation `param._saved_grad_shard.data += output.data` has conflicting devices.

rohan-varma replied:
When we are using gradient accumulation outside of no_sync + CPU offload, don't we already raise an appropriate error on L1336 of this PR? And is it possible to add a unittest for this?

awgu replied:
That is an internal assert. Given the current implementation, that assert will never get triggered. I added it because the logic is quite complicated, and I wanted to demonstrate that if we are in the non-`no_sync()` + CPU offloading case, then we should never be in that branch.

awgu replied:
(or more directly, the contrapositive: if we are in that branch, we should not be in the non-`no_sync()` + CPU offloading case, meaning that in particular we must not be CPU offloading)

rohan-varma reviewed (commented):
Thanks for turning this around so quickly! A couple of small comments, will stamp after those are addressed!

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.

This also adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.

This also adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.

This also adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.

This also adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.

This also adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1305,7 +1305,24 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                     # Average grad by world_size for consistency with PyTorch DDP.
                     output.div_(self.gradient_postdivide_factor)
-                param.grad.data = output
+                # To support gradient accumulation without `no_sync()`, we save
```
looks like it is important to do the following before flatten param.grad and pass it to reduce_scatter_base, as commented in https://github.com/facebookresearch/fairscale/commit/4fa2ab9b70d55f4e3e94239465b877c9f3733037

grad = param.grad.data
param.grad = None

This is to avoid race among multiple re-entrant backward passes.  for multiple re-entrant backward passes, e.g., it is possible that the second backward pass computation is ahead of first backward pass reduction, as reduction is in another stream and is async. In this case, first backward pass may be reducing the second pass's param.grad, which is not correct. 

so it is better to reduce_scatter the saved local grad and accumulate the reduced grad to saved_grad_shard. And then in the final callback of main backward pass, let param.grad point to saved_grad_shard. 

let's add change here and add comments as well

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1332,6 +1349,9 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                 # the transfer is async so it is not necessarily done until we
                 # explicitly synchronize in backward.
                 param.grad.data = param._cpu_grad  # type: ignore[attr-defined]
+                # We ensure that `param._saved_grad_shard` is on the same device
+                # as `param.grad` (to appease autograd).
+                param._saved_grad_shard = param._cpu_grad  # type: ignore[attr-defined]
```
based on above comments, we should move this to final callback of main backward pass

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1485,6 +1519,17 @@ def _prep_grads_for_backward(self) -> None:
+                    assert self.cpu_offload.offload_params, \
+                        "`p.grad.device` and `p.device` should be the same " \
+                        "if not offloading parameters to CPU"
+                can_accumulate_grad = p.grad.device == p.device and \
```
this means that currently grad will not be accumulated when cpu offloading is enabled? if so, let's add a comment

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1293,10 +1293,18 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
             if param._is_sharded:  # type: ignore[attr-defined]
-                grad_flatten = torch.flatten(param.grad)
+                # We clear `param.grad` to permit repeated gradient
+                # computations when this FSDP module is called multiple times.
```
could we add a comment to explain why the race could happen if not clear param.grad. "This is to avoid race among multiple re-entrant backward passes. for multiple re-entrant backward passes, e.g., it is possible that the second backward pass computation is ahead of first backward pass reduction, as reduction is in another stream and is async. In this case, first backward pass may be reducing the second pass's param.grad, which is not correct."

zhaojuanmao reviewed (commented):
fsdp code change looks good to me, will let @Rohan to take further look at the test codes. 

also, since this is nontrivial change, let's run A/B testing for fsdp accuracy and qps benchmarks as well

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1293,10 +1293,24 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
-                grad_flatten = torch.flatten(param.grad)
+                # We clear `param.grad` to permit repeated gradient
+                # computations when this FSDP module is called multiple times.
+                # This is to avoid a race among multiple re-entrant backward
```
is there a unittest that can capture this race condition?

awgu replied:
After offline discussion, we decided to omit such a unit test for now. If there is a non-contrived re-entrant backward use case for FSDP, we can revisit.

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1451,11 +1482,41 @@ def _remove_shard_bwd_hook(fsdp_module: FullyShardedDataParallel) -> None:
+                            f"p._saved_grad_shard={p._saved_grad_shard.device}"
+                        )
+                        p.grad = p._saved_grad_shard  # type: ignore[attr-defined]
+                    else:
+                        p_assert(
+                            not p._is_sharded, "All sharded parameters should "
+                            "use `_saved_grad_shard`"
+                        )
```
when users set p.grad=None, p will not have _saved_grad_shard, right?

awgu replied:
Yes, I believe this is the case.

`p._saved_grad_shard` is only set if `p.grad is not None` in `_prep_grads_for_backward()`.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1546,6 +1607,24 @@ def _prep_grads_for_backward(self) -> None:
+                        "if not offloading parameters to CPU"
+                outside_no_sync: bool = p.grad.size() == p._local_shard.shape  # type: ignore[attr-defined]
+                assert not outside_no_sync or \
+                    (self._is_root or self._require_backward_grad_sync), \
+                    "Each FSDP modules should require backward grad sync " \
+                    "outside `no_sync()` except the root module"
```
curious why except the root module?

awgu replied:
This is a great catch. I conflated two conditions together incorrectly.

I have a clearer version now.

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

rohan-varma commented on test/distributed/fsdp/test_fsdp_grad_acc.py:
```diff
@@ -0,0 +1,263 @@
+        )
+
+
+class TestGradAcc(FSDPTest):
```
Can we get a general idea of how much these tests increase TTS for OSS CI? For example I wonder if it is costly enough, should we consider running some of these tests on master-only.

awgu replied:
Updated version with only the two interleaved tests takes ~60 seconds on the AWS cluster.

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1546,6 +1608,28 @@ def _prep_grads_for_backward(self) -> None:
+                # try to accumulate gradients. FSDP accumulates gradients in
+                # the separate variable `p._saved_grad_shard` to leave `p.grad`
+                # for the per-iteration gradient.
+                if prev_iter_outside_no_sync:
```
Since this PR is getting a bit cluttered, I wanted to specifically point this part out. Previously, I presented the logic here incorrectly, but hopefully this should make sense now.

I think the precise condition of when to use `p._saved_grad_shard` is if the previous iteration was outside `no_sync()`.

Suppose we have
(1) some iterations outside `no_sync()` ->
(2) some iterations inside `no_sync()` ->
(3) one iteration outside `no_sync()`. 

- In the pre-backward hook of (3), the FSDP instance holds an unsharded gradient in `p.grad`, which is the result of accumulating gradients from (2).
- It computes that iteration's gradient, which is accumulated with the existing `p.grad` from (2) via the autograd engine and still stored in `p.grad`.
- In the post-backward hook, it reduce-scatters that accumulated gradient stored in `p.grad`.
- After the reduce-scatter, it accumulates the accumulated gradient from (2) and (3) with the accumulated gradient from (1) saved in `_saved_grad_shard`. This "super-accumulated" gradient is stored in `_saved_grad_shard`.
    - This step shows why as long as the previous iteration was outside `no_sync()`, there may be a gradient to accumulate on the first future iteration also outside `no_sync()`.

I had misunderstood the comments from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/40e7450f6fd1ba27a71c58da2b49cd19a46b7678/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L2000-L2003)). I did not realize that the conditioning on inside/outside `no_sync()` referred to the **previous** iteration.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1366,7 +1380,29 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
+                        f"existing grad shape={param._saved_grad_shard.shape} "
+                        f"new grad shape={output.shape}"  # type: ignore[attr-defined]
+                    )
+                    p_assert(
```
One more thing to point out: I previously had an assert like `not self.cpu_offload.offload_params` just as an internal assert to make sure that CPU offloading never takes this code path.

However, I changed it to a more direct assert here in case we distinguish between offloading parameters and offloading gradients in the future before we solve gradient accumulation with CPU offloading.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1366,7 +1380,29 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
+                        f"existing grad device={param._saved_grad_shard.device} "
+                        f"new grad device={output.device}"  # type: ignore[attr-defined]
+                    )
+                    param._saved_grad_shard.data += output.data  # type: ignore[attr-defined]
```
will it work if removing the .data?

awgu replied:
Yup, it still works. I will remove the `.data` for both in this line.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1546,6 +1608,28 @@ def _prep_grads_for_backward(self) -> None:
+                # the separate variable `p._saved_grad_shard` to leave `p.grad`
+                # for the per-iteration gradient.
+                if prev_iter_outside_no_sync:
+                    # FSDP currently does not support gradient accumulation
+                    # outside `no_sync()` when using CPU offloading. Trying to
+                    # do so yields incorrect results since FSDP will use the
+                    # newly-reduced gradient instead of accumulating with any
+                    # existing gradient.
```
could we add a github issue to support grad accumulation in cpu offloading?

awgu replied:
Done: https://github.com/pytorch/pytorch/issues/73784

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1546,6 +1608,28 @@ def _prep_grads_for_backward(self) -> None:
+                    # newly-reduced gradient instead of accumulating with any
+                    # existing gradient.
+                    if not offloaded:
+                        p._saved_grad_shard = p.grad.data  # type: ignore[attr-defined]
```
could we add a warning for this case and warning the comment "# FSDP currently does not support gradient accumulation
                    # outside `no_sync()` when using CPU offloading. Trying to
                    # do so yields incorrect results since FSDP will use the
                    # newly-reduced gradient instead of accumulating with any
                    # existing gradient."

zhaojuanmao reviewed (approved):
solid tests! also left some minor comments

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation without `no_sync()` and also interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

Commit:
Update on "[FSDP] Add grad accumulation without `no_sync()`"


**Overview**
- This adds FSDP gradient accumulation without `no_sync()`, which comparatively has more network bandwidth demand but less GPU memory requirement per worker.
- This fixes a bug in the `no_sync()` testing, where the CPU offloading and backward prefetch arguments were not propagating to the `FullyShardedDataParallel` constructor.
- This adds `p_assert()` (taken from Fairscale), which prints the assert error message before raising the `AssertionError`. It is meant to be used when running in the autograd backward context since otherwise the error message is swallowed, giving a unhelpful error like:
```
<built-in method run_backward of torch._C._EngineBase object at 0x7f1fd518dc80> returned NULL without setting an error
```

NOTE: Gradient accumulation without `no_sync()` is not currently compatible with CPU offloading.

**Test Plan**
I augmented the tests to test gradient accumulation interleaving iterations accumulating with and without `no_sync()`.

Differential Revision: [D34533546](https://our.internmc.facebook.com/intern/diff/D34533546)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34533546).

rohan-varma commented:
@awgu I was thinking maybe we should add some documentation somewhere around the gradient accumulation that is supported in FSDP, how to use it and what are the tradeoffs?

awgu commented:
> @awgu I was thinking maybe we should add some documentation somewhere around the gradient accumulation that is supported in FSDP, how to use it and what are the tradeoffs?

I think this is a great idea, especially since right now I do not have good intuition for the tradeoffs either.

rohan-varma commented:
@awgu Awesome! Feel free to file an issue so we don't forget about it:)

rohan-varma commented:
https://github.com/pytorch/pytorch/issues/74153
